crates/accelerate/src/circuit_library/entanglement.rs

```diff
@@ -0,0 +1,197 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use itertools::Itertools;
+use pyo3::{
+    types::{PyAnyMethods, PyInt, PyList, PyListMethods, PyString, PyTuple},
+    Bound, PyAny, PyResult,
+};
+
+use crate::QiskitError;
+
+/// Get all-to-all entanglement. For 4 qubits and block size 2 we have:
+/// [(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]
+pub fn full(num_qubits: u32, block_size: u32) -> impl Iterator<Item = Vec<u32>> {
+    (0..num_qubits).combinations(block_size as usize)
+}
+
+/// Get a linear entanglement structure. For ``n`` qubits and block size ``m`` we have:
+/// [(0..m-1), (1..m), (2..m+1), ..., (n-m..n-1)]
+pub fn linear(num_qubits: u32, block_size: u32) -> impl DoubleEndedIterator<Item = Vec<u32>> {
+    (0..num_qubits - block_size + 1)
+        .map(move |start_index| (start_index..start_index + block_size).collect())
+}
+
+/// Get a reversed linear entanglement. This is like linear entanglement but in reversed order:
+/// [(n-m..n-1), ..., (1..m), (0..m-1)]
+/// This is particularly interesting, as CX+"full" equals CX+"reverse_linear", even though the
+/// latter uses a quadratically less number of gates.
+pub fn reverse_linear(num_qubits: u32, block_size: u32) -> impl Iterator<Item = Vec<u32>> {
+    linear(num_qubits, block_size).rev()
+}
+
+/// Return the qubit indices for circular entanglement. This is defined as tuples of length ``m``
+/// starting at each possible index ``(0..n)``. Historically, Qiskit starts with index ``n-m+1``.
```

Comment:
is there a reason or reference for choosing this offset?

Reply:
I don't remember the details, though it might have been connected to the shifted-circular-alternating below... I added this as comment to explicitly state this behavior, which was uncommented before 🙂